clients/arweave_ipfs_s3.go

```diff
@@ -2,52 +2,96 @@ package clients
+func CopyDStorageToS3(url, s3URL string, requestID string) error {
+	content, err := DownloadDStorageFromGatewayList(url, requestID)
+	if err != nil {
+		return err
+	}
 
-func newDefaultRetryableHttpClient() *http.Client {
-	client := retryablehttp.NewClient()
-	client.RetryMax = 2                          // Retry a maximum of this+1 times
-	client.RetryWaitMin = 200 * time.Millisecond // Wait at least this long between retries
-	client.RetryWaitMax = 1 * time.Second        // Wait at most this long between retries (exponential backoff)
-	client.HTTPClient = &http.Client{
-		// Give up on requests that take more than this long - the file is probably too big for us to process locally if it takes this long
-		// or something else has gone wrong and the request is hanging
-		Timeout: MAX_COPY_DURATION,
+	err = UploadToOSURL(s3URL, "", content, MAX_COPY_DURATION)
+	if err != nil {
+		return err
 	}
 
-	return client.StandardClient()
+	return nil
 }
 
-func CopyArweaveToS3(arweaveURL, s3URL string) error {
-	resp, err := defaultRetryableHttpClient.Get(arweaveURL)
+func DownloadDStorageFromGatewayList(u string, requestID string) (io.ReadCloser, error) {
+	var err error
+	var gateways []*url.URL
+	dStorageURL, err := url.Parse(u)
+	if err != nil {
+		return nil, fmt.Errorf("invalid URL: %w", err)
+	}
+
+	switch dStorageURL.Scheme {
+	case SCHEME_ARWEAVE:
+		gateways = config.ImportArweaveGatewayURLs
+	case SCHEME_IPFS:
+		gateways = config.ImportIPFSGatewayURLs
+	default:
+		return nil, fmt.Errorf("unsupported dStorage scheme %s", dStorageURL.Scheme)
+	}
+
+	var opContent io.ReadCloser
+	downloadOperation := func() error {
+		for _, gateway := range gateways {
+			opContent = downloadDStorageResourceFromSingleGateway(gateway, dStorageURL.Host, requestID)
+			if opContent != nil {
+				return nil
+			}
+		}
+
+		return fmt.Errorf("failed to fetch %s from any of the gateways", u)
+	}
+
+	retryStrategy := backoff.NewConstantBackOff(1 * time.Second)
+	err = backoff.Retry(downloadOperation, backoff.WithMaxRetries(retryStrategy, 2))
 	if err != nil {
-		return fmt.Errorf("error fetching Arweave or IPFS URL: %s", err)
+		return nil, err
+	} else {
+		return opContent, nil
 	}
-	defer resp.Body.Close()
+}
+
+func downloadDStorageResourceFromSingleGateway(gateway *url.URL, cid, requestID string) io.ReadCloser {
+	fullURL := gateway.JoinPath(cid).String()
+	resp, err := http.DefaultClient.Get(fullURL)
 
-	err = UploadToOSURL(s3URL, "", resp.Body, MAX_COPY_DURATION)
 	if err != nil {
-		return fmt.Errorf("failed to copy Arweave or IPFS URL to S3: %s", err)
+		log.LogError(requestID, "failed to fetch content from gateway", err, "url", fullURL)
+		return nil
 	}
 
-	return nil
+	if resp.StatusCode >= 300 {
+		resp.Body.Close()
+		log.Log(requestID, "unexpected response from gateway", "status_code", resp.StatusCode, "url", fullURL)
+		return nil
+	}
+
+	return resp.Body
 }
 
-func IsArweaveOrIPFSURL(arweaveOrIPFSURL string) bool {
```

Comment:
Hmmm I just noticed a regression here in case the input is a URL to an IPFS/Arweave gateway. On that case, we also need to copy the file to an S3 bucket in order to get proper `Content-Length`, range requests, etc. If we don't copy these files from their gateways to an S3 bucket, the Mist processing *will* fail (which was the reason to create this copying in the first place)

We could either keep this "gateway URL identification" logic and replace it with the protocol-specififc scheme (like `ipfs://` or `ar://`), but that would potentially make the downloads slower or even not work at all (assuming the gateway provided by the user is the best gateway, possibly where the file resides).

So I think we need to either:
 - Change this function to also detect the old gateway URLs as DStorage URLs (then it might need to make the actual download function support these too), or;
 - Create a new function like `IsDStorageGatewayURL` which is used only on the place that used `IsArweaveOrIPFSURL` to copy it to an S3 before processing (in the Mist pipeline)

WDYT?

Reply:
Hmm, something more sophisticated but more overengineered would be to use the user-provided gateway as the _first_ gateway that we try, but still try the others. Not sure if it's worth the complexity tho, maybe better to just treat as a regular HTTPS URL for the download, but copy it to the bucket due to the gateway HTTP restritions.

Reply:
Good catch! Will push a solution